node/core/pvf/src/worker_common.rs

```diff
@@ -209,74 +198,55 @@ where
 		debug_id,
 		err,
 	);
+
+	// We don't want tokio to wait for the tasks to finish. We want to bring down the worker as fast
+	// as possible and not wait for stalled validation to finish. This isn't strictly necessary now,
+	// but may be in the future.
+	rt.shutdown_background();
 }
 
 /// Loop that runs in the CPU time monitor thread on prepare and execute jobs. Continuously wakes up
-/// from sleeping and then either sleeps for the remaining CPU time, or sends back a timeout error
-/// if we exceed the CPU timeout.
+/// from sleeping and then either sleeps for the remaining CPU time, or returns if we exceed the CPU
+/// timeout.
+///
+/// Returning indicates that we should send a `TimedOut` error to the host.
 ///
-/// NOTE: If the job completes and this thread is still sleeping, it will continue sleeping in the
-/// background. When it wakes, it will see that the flag has been set and return.
-pub async fn cpu_time_monitor_loop(
+/// NOTE: Returning will cause the worker, whether preparation or execution, to be killed by the
+/// host. We do not kill the process here because it would interfere with the proper handling of
+/// this error.
+pub fn cpu_time_monitor_loop(
 	job_kind: JobKind,
-	mut stream: UnixStream,
 	cpu_time_start: ProcessTime,
 	timeout: Duration,
-	lock: Arc<AtomicBool>,
+	finished_rx: Receiver<()>,
 ) {
 	loop {
 		let cpu_time_elapsed = cpu_time_start.elapsed();
 
 		// Treat the timeout as CPU time, which is less subject to variance due to load.
-		if cpu_time_elapsed > timeout {
```

Comment:
You should still check if `finished.try_recv().map(|_| true).unwrap_or_else(|e| e == TryRecvError::Disconnected)`

Otherwise we may print this warning below, which would not be true.

Reply:
I see the issue. Your change wouldn't fix it though, because the job (other thread) can still finish after we log. It can technically even finish between the `try_recv()` check and the `map`.

Maybe best would be to move this to the `select!`s, where we are guaranteed to have chosen the timed-out branch, although there would be a bit of duplication.

Also, I think it's possible for this thread to complete before we execute `finished_tx.send`, which would trigger an error as the receiver has already been dropped. And right now, such a spurious error from `send` would be returned even if the job otherwise succeeded.

I think I ran into just about every async edge case in this PR. 😛 I picked up this issue to practice async programming and got my wish I guess!

Reply:
Good point!